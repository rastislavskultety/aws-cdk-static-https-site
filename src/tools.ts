
/**
 * Simple object check.
 * @param item
 * @returns {boolean}
 */
export function isObject(item: any): boolean {
  return (item && typeof item === 'object' && !Array.isArray(item));
}

/**
 * Deep merge two objects.
 * @param target
 * @param ...sources
 */
export function mergeDeep(target: any, ...sources: any[]): any {
  if (!sources.length) return target;
  const source = sources.shift();

  if (isObject(target) && isObject(source)) {
    for (const key in source) {
      if (isObject(source[key])) {
        if (!isObject(target[key])) Object.assign(target, { [key]: {} });
        mergeDeep(target[key], source[key]);
      } else {
        Object.assign(target, { [key]: source[key] });
      }
    }
  }

  return mergeDeep(target, ...sources);
}

/**
 * Helper to set default properties or override existing properties.
 * @param props - current properties
 * @param propsDefaults - defaults and overrides
 * @param deepMerge - when true then nested objects are merged as well, otherways only the root level is merged.
 * @returns properties with merged defaults and overrides.
 */
export function makeProps<T>(props: T, propsDefaults?: { defaults?: any; overrides?: any }, deepMerge: boolean = false): T {
  if (propsDefaults) {
    const fn = deepMerge ? mergeDeep : Object.assign;
    return fn({}, propsDefaults.defaults, props, propsDefaults.overrides);
  }
  return props;
}

/**
 * Logical exclusive or function
 * @param a - first parameter
 * @param b - second parameter
 * @returns a exclusive or b
 */
export function xor(a: any, b: any): boolean {
  return !!(a ? !b : b);
}
