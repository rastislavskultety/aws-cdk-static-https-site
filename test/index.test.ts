import { SynthUtils } from '@aws-cdk/assert';
import { Stack } from '@aws-cdk/core';
import { StaticWebSiteProps, StaticWebSite, StaticWebSitePrimaryDomain } from '../src';
import '@aws-cdk/assert/jest';

function createStack() {
  return new Stack(undefined, undefined, {
    env: {
      account: 'accountid',
      region: 'us-east-1',
    },
  });
}

const defaultProps = {
  rootDomain: 'example.org',
  siteSubDomain: 'www',
  siteContentsPath: './test/test-website',
  useRoute53: true,
  redirectSecondaryDomain: true,
};

function getProps(customProps?: object): StaticWebSiteProps {
  return Object.assign({}, defaultProps, customProps) as StaticWebSiteProps;
}

function createTestCase(customProps?: object): { stack: Stack; site: StaticWebSite } {
  const stack = createStack();
  const site = new StaticWebSite(stack, 'test', getProps(customProps));
  return { stack, site };
}

test('Refactoring test using snapshot', () => {
  const { stack } = createTestCase();
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

test('All resources are created', () => {
  const { stack } = createTestCase();

  expect(stack).toHaveResourceLike('AWS::S3::Bucket', { BucketName: 'www.example.org' });
  expect(stack).toHaveResourceLike('AWS::S3::Bucket', { BucketName: 'example.org' });
  expect(stack).toHaveResourceLike('AWS::CloudFront::Distribution', {
    DistributionConfig: { Aliases: ['www.example.org'] },
  });
  expect(stack).toHaveResourceLike('AWS::CloudFront::Distribution', {
    DistributionConfig: { Aliases: ['example.org'] },
  });
  expect(stack).toHaveResourceLike('AWS::CertificateManager::Certificate', {
    DomainName: 'www.example.org',
  });
  expect(stack).toHaveResource('AWS::Route53::RecordSet', { Name: 'example.org.', Type: 'A' });
  expect(stack).toHaveResource('AWS::Route53::RecordSet', { Name: 'www.example.org.', Type: 'A' });
});

test('Domain names are correct for primaryDomain=SUB_DOMAIN', () => {
  const { site } = createTestCase({ primaryDomain: StaticWebSitePrimaryDomain.SUB_DOMAIN });

  expect(site.siteDomain).toBe('www.example.org');
  expect(site.redirectedDomain).toBe('example.org');
});

test('Domain names are correct for primaryDomain=ROOT_DOMAIN', () => {
  const { site } = createTestCase({ primaryDomain: StaticWebSitePrimaryDomain.ROOT_DOMAIN });
  expect(site.siteDomain).toBe('example.org');
  expect(site.redirectedDomain).toBe('www.example.org');
});

test('Creating Wildcard certificate', () => {
  const { stack } = createTestCase({ createWildcardCertificate: true });
  expect(stack).toHaveResourceLike('AWS::CertificateManager::Certificate', {
    DomainName: '*.example.org',
  });
});

test('Default index document', () => {
  const { stack } = createTestCase();

  expect(stack).toHaveResource('AWS::S3::Bucket', {
    BucketName: 'www.example.org',
    WebsiteConfiguration: {
      IndexDocument: 'index.html',
    },
  });
});

test('Specific index and error document', () => {
  const { stack } = createTestCase({
    websiteIndexDocument: 'main.html',
    websiteErrorDocument: '404.html',
  });
  expect(stack).toHaveResource('AWS::S3::Bucket', {
    BucketName: 'www.example.org',
    WebsiteConfiguration: {
      ErrorDocument: '404.html',
      IndexDocument: 'main.html',
    },
  });
});

test('Using root domain only for the web site', () => {
  const { stack } = createTestCase({
    siteSubDomain: '',
    primaryDomain: StaticWebSitePrimaryDomain.ROOT_DOMAIN,
    redirectSecondaryDomain: false,
  });
  expect(stack).toHaveResourceLike('AWS::S3::Bucket', { BucketName: 'example.org' });
});

test('Checking invalid properties', () => {
  // Root domain must be always specified.
  expect(() => createTestCase({ rootDomain: null })).toThrowError(/rootDomain must be specified/);

  // Site contents must be always specified.
  expect(() => createTestCase({ siteContentsPath: null })).toThrowError(/siteContentsPath must be specified/);

  // Sub domain must be specified if it is primary
  expect(() => createTestCase({ siteSubDomain: '' })).toThrowError(/siteSubDomain must be specified/);

  // Redirection is not valid if there is no sub domain
  expect(() => createTestCase({
    siteSubDomain: '',
    primaryDomain: StaticWebSitePrimaryDomain.ROOT_DOMAIN,
    redirectSecondaryDomain: true,
  })).toThrowError(/siteSubDomain must be specified/);

  // Cannot use wildcard for certificate if primary domain = ROOT_DOMAIN
  expect(() => createTestCase({
    createWildcardCertificate: true,
    primaryDomain: StaticWebSitePrimaryDomain.ROOT_DOMAIN,
  })).toThrowError(/createWildcardCertificate/);

  // Certificate ARN must be specified if not using Route 53
  expect(() => createTestCase({ useRoute53: false })).toThrowError(/certificateArn or userRoute53/);

  // Cannot use redirection if not using route53
  expect(() => createTestCase({
    useRoute53: false,
    certificateArn: 'some',
    primaryDomain: StaticWebSitePrimaryDomain.ROOT_DOMAIN,
  })).toThrowError(/root domain cannot be used/);

  // Cannot use redirection if not using route53
  expect(() => createTestCase({ useRoute53: false, certificateArn: 'some' })).toThrowError(/redirection .* not possible/);

});
