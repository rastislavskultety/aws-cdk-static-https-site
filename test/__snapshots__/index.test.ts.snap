// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Refactoring test using snapshot 1`] = `
Object {
  "Mappings": Object {
    "AWSCloudFrontPartitionHostedZoneIdMap": Object {
      "aws": Object {
        "zoneId": "Z2FDTNDATAQYW2",
      },
      "aws-cn": Object {
        "zoneId": "Z3RFFRIM2A3IF5",
      },
    },
  },
  "Outputs": Object {
    "testRedirectedUrl3549E758": Object {
      "Value": "https://example.org",
    },
    "testSiteUrl6033360F": Object {
      "Value": "https://www.example.org",
    },
  },
  "Parameters": Object {
    "AssetParameters4cd61014b71160e8c66fe167e43710d5ba068b80b134e9bd84508cf9238b2392ArtifactHashE56CD69A": Object {
      "Description": "Artifact hash for asset \\"4cd61014b71160e8c66fe167e43710d5ba068b80b134e9bd84508cf9238b2392\\"",
      "Type": "String",
    },
    "AssetParameters4cd61014b71160e8c66fe167e43710d5ba068b80b134e9bd84508cf9238b2392S3BucketBF7A7F3F": Object {
      "Description": "S3 bucket for asset \\"4cd61014b71160e8c66fe167e43710d5ba068b80b134e9bd84508cf9238b2392\\"",
      "Type": "String",
    },
    "AssetParameters4cd61014b71160e8c66fe167e43710d5ba068b80b134e9bd84508cf9238b2392S3VersionKeyFAF93626": Object {
      "Description": "S3 key for asset version \\"4cd61014b71160e8c66fe167e43710d5ba068b80b134e9bd84508cf9238b2392\\"",
      "Type": "String",
    },
    "AssetParametersa189a6dc4eac353ddaccaf24d3da55780af0794e2deb290c6d87be21501ccbbfArtifactHashAED8DFE5": Object {
      "Description": "Artifact hash for asset \\"a189a6dc4eac353ddaccaf24d3da55780af0794e2deb290c6d87be21501ccbbf\\"",
      "Type": "String",
    },
    "AssetParametersa189a6dc4eac353ddaccaf24d3da55780af0794e2deb290c6d87be21501ccbbfS3Bucket2890265E": Object {
      "Description": "S3 bucket for asset \\"a189a6dc4eac353ddaccaf24d3da55780af0794e2deb290c6d87be21501ccbbf\\"",
      "Type": "String",
    },
    "AssetParametersa189a6dc4eac353ddaccaf24d3da55780af0794e2deb290c6d87be21501ccbbfS3VersionKeyA79804AD": Object {
      "Description": "S3 key for asset version \\"a189a6dc4eac353ddaccaf24d3da55780af0794e2deb290c6d87be21501ccbbf\\"",
      "Type": "String",
    },
    "AssetParametersc24b999656e4fe6c609c31bae56a1cf4717a405619c3aa6ba1bc686b8c2c86cfArtifactHash85F58E48": Object {
      "Description": "Artifact hash for asset \\"c24b999656e4fe6c609c31bae56a1cf4717a405619c3aa6ba1bc686b8c2c86cf\\"",
      "Type": "String",
    },
    "AssetParametersc24b999656e4fe6c609c31bae56a1cf4717a405619c3aa6ba1bc686b8c2c86cfS3Bucket55EFA30C": Object {
      "Description": "S3 bucket for asset \\"c24b999656e4fe6c609c31bae56a1cf4717a405619c3aa6ba1bc686b8c2c86cf\\"",
      "Type": "String",
    },
    "AssetParametersc24b999656e4fe6c609c31bae56a1cf4717a405619c3aa6ba1bc686b8c2c86cfS3VersionKey60329B70": Object {
      "Description": "S3 key for asset version \\"c24b999656e4fe6c609c31bae56a1cf4717a405619c3aa6ba1bc686b8c2c86cf\\"",
      "Type": "String",
    },
    "AssetParameterse9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68ArtifactHashD9A515C3": Object {
      "Description": "Artifact hash for asset \\"e9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68\\"",
      "Type": "String",
    },
    "AssetParameterse9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68S3BucketAEADE8C7": Object {
      "Description": "S3 bucket for asset \\"e9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68\\"",
      "Type": "String",
    },
    "AssetParameterse9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68S3VersionKeyE415415F": Object {
      "Description": "S3 key for asset version \\"e9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536": Object {
      "DependsOn": Array [
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersc24b999656e4fe6c609c31bae56a1cf4717a405619c3aa6ba1bc686b8c2c86cfS3Bucket55EFA30C",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersc24b999656e4fe6c609c31bae56a1cf4717a405619c3aa6ba1bc686b8c2c86cfS3VersionKey60329B70",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersc24b999656e4fe6c609c31bae56a1cf4717a405619c3aa6ba1bc686b8c2c86cfS3VersionKey60329B70",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.handler",
        "Layers": Array [
          Object {
            "Ref": "testDeployWithInvalidationAwsCliLayer350135E8",
          },
        ],
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
            "Arn",
          ],
        },
        "Runtime": "python3.6",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "AssetParametersa189a6dc4eac353ddaccaf24d3da55780af0794e2deb290c6d87be21501ccbbfS3Bucket2890265E",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "AssetParametersa189a6dc4eac353ddaccaf24d3da55780af0794e2deb290c6d87be21501ccbbfS3Bucket2890265E",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "testSiteBucket233E33D9",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "testSiteBucket233E33D9",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "cloudfront:GetInvalidation",
                "cloudfront:CreateInvalidation",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "Roles": Array [
          Object {
            "Ref": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F": Object {
      "DependsOn": Array [
        "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters4cd61014b71160e8c66fe167e43710d5ba068b80b134e9bd84508cf9238b2392S3BucketBF7A7F3F",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters4cd61014b71160e8c66fe167e43710d5ba068b80b134e9bd84508cf9238b2392S3VersionKeyFAF93626",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters4cd61014b71160e8c66fe167e43710d5ba068b80b134e9bd84508cf9238b2392S3VersionKeyFAF93626",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Description": Object {
          "Fn::Join": Array [
            "",
            Array [
              "Lambda function for auto-deleting objects in ",
              Object {
                "Ref": "testSiteBucket233E33D9",
              },
              " S3 bucket.",
            ],
          ],
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "testCertificate6A3C712A": Object {
      "Properties": Object {
        "DomainName": "www.example.org",
        "DomainValidationOptions": Array [
          Object {
            "DomainName": "www.example.org",
            "HostedZoneId": "DUMMY",
          },
          Object {
            "DomainName": "example.org",
            "HostedZoneId": "DUMMY",
          },
        ],
        "SubjectAlternativeNames": Array [
          "example.org",
        ],
        "ValidationMethod": "DNS",
      },
      "Type": "AWS::CertificateManager::Certificate",
    },
    "testDeployWithInvalidationAwsCliLayer350135E8": Object {
      "Properties": Object {
        "Content": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameterse9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68S3BucketAEADE8C7",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameterse9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68S3VersionKeyE415415F",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameterse9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68S3VersionKeyE415415F",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Description": "/opt/awscli/aws",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "testDeployWithInvalidationCustomResourceE35F32E3": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "DestinationBucketName": Object {
          "Ref": "testSiteBucket233E33D9",
        },
        "DistributionId": Object {
          "Ref": "testSiteDistributionCFDistribution9443B2CC",
        },
        "DistributionPaths": Array [
          "/*",
        ],
        "Prune": true,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536",
            "Arn",
          ],
        },
        "SourceBucketNames": Array [
          Object {
            "Ref": "AssetParametersa189a6dc4eac353ddaccaf24d3da55780af0794e2deb290c6d87be21501ccbbfS3Bucket2890265E",
          },
        ],
        "SourceObjectKeys": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersa189a6dc4eac353ddaccaf24d3da55780af0794e2deb290c6d87be21501ccbbfS3VersionKeyA79804AD",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersa189a6dc4eac353ddaccaf24d3da55780af0794e2deb290c6d87be21501ccbbfS3VersionKeyA79804AD",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        ],
      },
      "Type": "Custom::CDKBucketDeployment",
      "UpdateReplacePolicy": "Delete",
    },
    "testRedirectionAliasRecordD9513E00": Object {
      "Properties": Object {
        "AliasTarget": Object {
          "DNSName": Object {
            "Fn::GetAtt": Array [
              "testRedirectionDistributionCFDistribution36DFE729",
              "DomainName",
            ],
          },
          "HostedZoneId": Object {
            "Fn::FindInMap": Array [
              "AWSCloudFrontPartitionHostedZoneIdMap",
              Object {
                "Ref": "AWS::Partition",
              },
              "zoneId",
            ],
          },
        },
        "HostedZoneId": "DUMMY",
        "Name": "example.org.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "testRedirectionBucketF7AFF040": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "BucketName": "example.org",
        "WebsiteConfiguration": Object {
          "RedirectAllRequestsTo": Object {
            "HostName": "www.example.org",
            "Protocol": "https",
          },
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "testRedirectionBucketPolicy3E607965": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "testRedirectionBucketF7AFF040",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "testRedirectionBucketF7AFF040",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "testRedirectionDistributionCFDistribution36DFE729": Object {
      "Properties": Object {
        "DistributionConfig": Object {
          "Aliases": Array [
            "example.org",
          ],
          "DefaultCacheBehavior": Object {
            "AllowedMethods": Array [
              "GET",
              "HEAD",
            ],
            "CachedMethods": Array [
              "GET",
              "HEAD",
            ],
            "Compress": true,
            "ForwardedValues": Object {
              "Cookies": Object {
                "Forward": "none",
              },
              "QueryString": false,
            },
            "TargetOriginId": "origin1",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": Array [
            Object {
              "ConnectionAttempts": 3,
              "ConnectionTimeout": 10,
              "CustomOriginConfig": Object {
                "HTTPPort": 80,
                "HTTPSPort": 443,
                "OriginKeepaliveTimeout": 5,
                "OriginProtocolPolicy": "http-only",
                "OriginReadTimeout": 30,
                "OriginSSLProtocols": Array [
                  "TLSv1.2",
                ],
              },
              "DomainName": Object {
                "Fn::Select": Array [
                  2,
                  Object {
                    "Fn::Split": Array [
                      "/",
                      Object {
                        "Fn::GetAtt": Array [
                          "testRedirectionBucketF7AFF040",
                          "WebsiteURL",
                        ],
                      },
                    ],
                  },
                ],
              },
              "Id": "origin1",
            },
          ],
          "PriceClass": "PriceClass_100",
          "ViewerCertificate": Object {
            "AcmCertificateArn": Object {
              "Ref": "testCertificate6A3C712A",
            },
            "MinimumProtocolVersion": "TLSv1.1_2016",
            "SslSupportMethod": "sni-only",
          },
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "testSiteAliasRecordA152B6BB": Object {
      "Properties": Object {
        "AliasTarget": Object {
          "DNSName": Object {
            "Fn::GetAtt": Array [
              "testSiteDistributionCFDistribution9443B2CC",
              "DomainName",
            ],
          },
          "HostedZoneId": Object {
            "Fn::FindInMap": Array [
              "AWSCloudFrontPartitionHostedZoneIdMap",
              Object {
                "Ref": "AWS::Partition",
              },
              "zoneId",
            ],
          },
        },
        "HostedZoneId": "DUMMY",
        "Name": "www.example.org.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "testSiteBucket233E33D9": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "BucketName": "www.example.org",
        "WebsiteConfiguration": Object {
          "IndexDocument": "index.html",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "testSiteBucketAutoDeleteObjectsCustomResourceE63A9A20": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "testSiteBucketPolicy5D0A3228",
      ],
      "Properties": Object {
        "BucketName": Object {
          "Ref": "testSiteBucket233E33D9",
        },
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F",
            "Arn",
          ],
        },
      },
      "Type": "Custom::S3AutoDeleteObjects",
      "UpdateReplacePolicy": "Delete",
    },
    "testSiteBucketPolicy5D0A3228": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "testSiteBucket233E33D9",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "testSiteBucket233E33D9",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::GetAtt": Array [
                    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
                    "Arn",
                  ],
                },
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "testSiteBucket233E33D9",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "testSiteBucket233E33D9",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "testSiteDistributionCFDistribution9443B2CC": Object {
      "Properties": Object {
        "DistributionConfig": Object {
          "Aliases": Array [
            "www.example.org",
          ],
          "DefaultCacheBehavior": Object {
            "AllowedMethods": Array [
              "GET",
              "HEAD",
            ],
            "CachedMethods": Array [
              "GET",
              "HEAD",
            ],
            "Compress": true,
            "ForwardedValues": Object {
              "Cookies": Object {
                "Forward": "none",
              },
              "QueryString": false,
            },
            "TargetOriginId": "origin1",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": Array [
            Object {
              "ConnectionAttempts": 3,
              "ConnectionTimeout": 10,
              "CustomOriginConfig": Object {
                "HTTPPort": 80,
                "HTTPSPort": 443,
                "OriginKeepaliveTimeout": 5,
                "OriginProtocolPolicy": "http-only",
                "OriginReadTimeout": 30,
                "OriginSSLProtocols": Array [
                  "TLSv1.2",
                ],
              },
              "DomainName": Object {
                "Fn::Select": Array [
                  2,
                  Object {
                    "Fn::Split": Array [
                      "/",
                      Object {
                        "Fn::GetAtt": Array [
                          "testSiteBucket233E33D9",
                          "WebsiteURL",
                        ],
                      },
                    ],
                  },
                ],
              },
              "Id": "origin1",
            },
          ],
          "PriceClass": "PriceClass_100",
          "ViewerCertificate": Object {
            "AcmCertificateArn": Object {
              "Ref": "testCertificate6A3C712A",
            },
            "MinimumProtocolVersion": "TLSv1.1_2016",
            "SslSupportMethod": "sni-only",
          },
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
  },
}
`;
